Fix incorrect abi encoding of storage array of data type that occupy multiple storage slots

sifmelcara opened the pull request:
### Description

Please see the report for details.

Note that `uint256[][]` and `uint256[][2]` do not even compile at this moment, so they are not included in the test. I will include them in the test at #6881 . (So this PR and #6881 somewhat depends on each other)

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [x] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

chriseth commented on Changelog.md:
```diff
@@ -11,6 +11,7 @@ Compiler Features:
 Bugfixes:
  * Yul / Inline Assembly Parser: Disallow trailing commas in function call arguments.
  * Set state mutability of the function type members ``gas`` and ``value`` to pure (while their return type inherits state mutability from the function type).
+ * Fix incorrect abi encoding of storage array of data type that occupy multiple storage slots
```
Should move to its own section at the top titled `Important Bugfixes`

chriseth commented on libsolidity/codegen/YulUtilFunctions.cpp:
```diff
@@ -592,7 +592,7 @@ string YulUtilFunctions::nextArrayElementFunction(ArrayType const& _type)
 		if (_type.location() == DataLocation::Memory)
 			templ("advance", "0x20");
 		else if (_type.location() == DataLocation::Storage)
-			templ("advance", "1");
```
Please also add an assertion that it is `>= 1`.

sifmelcara replied:
There is already `solAssert(_type.baseType()->storageBytes() > 16, "")` in line number 583. Do we want to assert again?

sifmelcara replied:
(I mean, `storageBytes() > 16` somewhat implies `storageSize()` is not zero?)

chriseth replied:
Better safe than sorry! :)

sifmelcara replied:
Ok. I also add an assertion for `calldataEncodedSize()`

sifmelcara replied:
I hope `solAssert(size >= 32 && size % 32 == 0, "");` makes sense for calldata types. Looks like they always pad to word except `bytes` and `string`.

chriseth replied:
Yep, looks good!

chriseth commented:
Ah sorry, should have mentioned that earlier:

This needs an entry in the buglist (`doc/bugs.json`). Maybe something like the following:
```
    {
        "name": "StorageArrayABIEncoderV2",
        "summary": "Storage arrays containing structs or other arrays are not read properly when directly encoded in function calls or in abi.encode*.",
        "description": "When storage arrays whose elements occupy more than a single storage slot are directly encoded in function calls or using abi.encode*, their elements are read in an overlapping manner, i.e. the element pointer is not properly advanced between reads. This is not a problem when the storage data is first copied to a memory variable.",
        "introduced": "0.4.16",
        "fixed": "0.5.10",
        "severity": "low",
        "conditions": {
            "ABIEncoderV2": true
        }
    },

```

sifmelcara commented:
Seems like there are two entries (`ABIEncoderV2PackedStorage`, `ABIEncoderV2PackedStorage_0.4.x`) for the previous abiencoderv2 bug, do we also need to add two entries this time?

sifmelcara commented:
Oh I guess it depends on whether we will backport this bugfix to 0.4.x version

sifmelcara committed:
Fix storage array abi encoding

Fix incorrect abi encoding of storage array of data type that occupy multiple storage slots

sifmelcara committed:
Add test case that abi.encode nested storage arrays

sifmelcara committed:
Add ABIEncoderV2 storage array encoding bug to buglist
